File: solutions/php/nth-prime/1/NthPrime.php
<?php
declare(strict_types=1);

function prime(int $number)
{
    if ($number < 1) return false;   // ← return false, not throw

    $count = 0;
    $candidate = 2;
    while (true) {
        if (isPrime($candidate)) {
            $count++;
            if ($count === $number) return $candidate;
        }
        $candidate++;
    }
}

function isPrime(int $n): bool
{
    if ($n < 2) return false;
    if ($n % 2 === 0) return $n === 2;
    for ($d = 3; $d * $d <= $n; $d += 2) {
        if ($n % $d === 0) return false;
    }
    return true;
}

File: solutions/ruby/nth-prime/1/nth_prime.rb
class Prime
  def self.nth(n)
    # 1. Validamos la entrada
    raise ArgumentError, "n debe ser un entero positivo" if n < 1

    # 2. Manejamos el primer caso manualmente para simplificar el bucle
    return 2 if n == 1

    count = 1 # Ya contamos al 2
    candidate = 3 # Empezamos a probar desde el primer impar

    while count < n
      if prime?(candidate)
        count += 1
      end
      
      # Si ya llegamos al conteo, devolvemos el candidato antes de incrementarlo
      return candidate if count == n
      
      # Saltamos de 2 en 2 para probar solo números impares
      candidate += 2
    end
  end

  private

  def self.prime?(num)
    # Un número es primo si no es divisible por ningún impar hasta su raíz cuadrada
    limit = Math.sqrt(num).to_i
    (3..limit).step(2) do |i|
      return false if num % i == 0
    end
    true
  end
end